File: zero-server/src/main/kotlin/com/zeroday/service/HackToLearnService.kt
package com.zeroday.service

import com.zeroday.model.CommandRegistry
import com.zeroday.zdscript.NexusScript
import com.zeroday.zdscript.NexusScriptEngine
import com.zeroday.zdscript.ScriptContext
import com.zeroday.zdscript.NexusValue
import com.zeroday.zdscript.KnowledgeFragment
import com.zeroday.zdscript.KnowledgeCategory
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

class HackToLearnService(
    private val playerService: PlayerService,
    private val gameEventBus: GameEventBus? = null
) {
    private val engine = NexusScriptEngine()
    private val playerKnowledge = mutableMapOf<String, MutableSet<String>>()
    private val playerScripts = mutableMapOf<String, MutableList<NexusScript>>()
    private val playerDiscoveredCommands = mutableMapOf<String, MutableSet<String>>()
    private val mutex = Mutex()

    suspend fun initializePlayer(playerId: String) = mutex.withLock {
        playerKnowledge.getOrPut(playerId) { mutableSetOf() }
        playerScripts.getOrPut(playerId) { mutableListOf() }
        playerDiscoveredCommands.getOrPut(playerId) { mutableSetOf() }
    }

    suspend fun learnKnowledge(playerId: String, fragmentId: String): Boolean = mutex.withLock {
        val knowledge = playerKnowledge.getOrPut(playerId) { mutableSetOf() }
        knowledge.add(fragmentId)
    }

    suspend fun getPlayerKnowledge(playerId: String): Set<String> = mutex.withLock {
        playerKnowledge[playerId]?.toSet() ?: emptySet()
    }

    suspend fun validateScript(code: String): NexusScriptEngine.ValidationResult =
        engine.validate(code)

    suspend fun saveScript(playerId: String, name: String, code: String, description: String): Result<NexusScript> = mutex.withLock {
        val validation = engine.validate(code)
        if (!validation.valid) {
            return@withLock Result.failure(Exception("Script validation failed: ${validation.errors.joinToString("; ")}"))
        }
        val script = NexusScript(
            name = name,
            authorId = playerId,
            code = code,
            description = description,
            validated = true
        )
        playerScripts.getOrPut(playerId) { mutableListOf() }.add(script)
        if (gameEventBus != null) {
            gameEventBus.emit(GameEvent(playerId, com.zeroday.model.AchievementEvent.SCRIPT_SAVED, 1L))
        }
        Result.success(script)
    }

    suspend fun getPlayerScripts(playerId: String): List<NexusScript> = mutex.withLock {
        playerScripts[playerId]?.toList() ?: emptyList()
    }

    suspend fun deleteScript(playerId: String, scriptId: String): Result<Unit> = mutex.withLock {
        val removed = playerScripts[playerId]?.removeAll { it.id == scriptId } ?: false
        if (removed) Result.success(Unit)
        else Result.failure(Exception("Script not found"))
    }

    suspend fun discoverFragment(playerId: String, fragmentId: String): Result<KnowledgeFragment> = mutex.withLock {
        val fragment = knowledgeFragments.find { it.id == fragmentId }
            ?: return@withLock Result.failure(Exception("Knowledge fragment not found"))

        val known = playerKnowledge.getOrPut(playerId) { mutableSetOf() }
        if (fragmentId in known) {
            return@withLock Result.failure(Exception("Already discovered this fragment"))
        }
        if (!fragment.requiredFragments.all { it in known }) {
            return@withLock Result.failure(Exception("Required fragments: ${fragment.requiredFragments.joinToString(", ")}"))
        }
        known.add(fragmentId)

        fragment.unlocksCommand?.let { cmd ->
            playerDiscoveredCommands.getOrPut(playerId) { mutableSetOf() }.add(cmd)
            playerService.unlockCommand(playerId, cmd)
        }
        if (gameEventBus != null) {
            gameEventBus.emit(GameEvent(playerId, com.zeroday.model.AchievementEvent.FRAGMENT_GATHERED, 1L))
        }
        Result.success(fragment)
    }

    suspend fun getAvailableFragments(playerId: String): List<KnowledgeFragment> = mutex.withLock {
        val known = playerKnowledge.getOrPut(playerId) { mutableSetOf() }
        knowledgeFragments.filter { f ->
            f.id !in known && f.requiredFragments.all { it in known }
        }
    }

    suspend fun getKnowledgeMap(playerId: String): KnowledgeMapData = mutex.withLock {
        val known = playerKnowledge.getOrPut(playerId) { mutableSetOf() }
        val discoveredCmds = playerDiscoveredCommands.getOrPut(playerId) { mutableSetOf() }

        KnowledgeMapData(
            totalFragments = knowledgeFragments.size,
            discoveredFragments = known.size,
            unlockedCommands = discoveredCmds.size,
            totalCommands = CommandRegistry.allCommands.size,
            fragments = knowledgeFragments.map { f ->
                FragmentStatus(
                    id = f.id, name = f.name, description = f.description,
                    category = f.category.name, rarity = f.rarity,
                    unlocked = f.id in known,
                    available = f.id !in known && f.requiredFragments.all { it in known },
                    requiredFragments = f.requiredFragments,
                    unlocksCommand = f.unlocksCommand,
                    sourceHint = if (f.id in known) "" else f.sourceHint
                )
            }
        )
    }

    suspend fun runCode(playerId: String, code: String): Result<NexusScriptEngine.ScriptResult> = mutex.withLock {
        val validation = engine.validate(code)
        if (!validation.valid) {
            return@withLock Result.failure(Exception(validation.errors.joinToString("; ")))
        }
        val player = playerService.getPlayer(playerId)
            ?: return@withLock Result.failure(Exception("Player not found"))
        val context = ScriptContext(playerId, player.level)
        val script = NexusScript(name = "inline", authorId = playerId, code = code, validated = true)
        val result = engine.execute(script, context)
        if (result.success) Result.success(result)
        else Result.failure(Exception(result.error))
    }

    suspend fun executeScript(playerId: String, scriptId: String, scriptArgs: Map<String, String> = emptyMap()): Result<NexusScriptEngine.ScriptResult> = mutex.withLock {
        val script = playerScripts[playerId]?.find { it.id == scriptId }
            ?: return@withLock Result.failure(Exception("Script not found"))
        val player = playerService.getPlayer(playerId)
            ?: return@withLock Result.failure(Exception("Player not found"))
        val context = ScriptContext(playerId, player.level)
        scriptArgs.forEach { (k, v) -> context.variables[k] = NexusValue.StrVal(v) }
        val result = engine.execute(script, context)
        if (result.success) Result.success(result)
        else Result.failure(Exception(result.error))
    }

    companion object {
        private val knowledgeFragments: List<KnowledgeFragment> = listOf(
            KnowledgeFragment("recon_basics", "Recon Basics", "Foundational reconnaissance techniques",
                KnowledgeCategory.RECONNAISSANCE, listOf(), "ping", "common", "Run 'ping' or 'scan' to discover this"),
            KnowledgeFragment("recon_advanced", "Advanced Recon", "Deep network probing and enumeration",
                KnowledgeCategory.RECONNAISSANCE, listOf("recon_basics"), "nmap", "uncommon", "Run 'nmap' on a target"),
            KnowledgeFragment("exploit_basics", "Exploit Fundamentals", "Basic exploitation techniques",
                KnowledgeCategory.EXPLOITATION, listOf("recon_advanced"), "exploit", "rare", "Run 'exploit' on a target"),
            KnowledgeFragment("crypto_basics", "Crypto Foundations", "Understanding encryption and hashing",
                KnowledgeCategory.CRYPTOGRAPHY, listOf(), "decrypt", "common", "Run 'decrypt' on a file"),
            KnowledgeFragment("crypto_advanced", "Crypto Mastery", "Breaking strong encryption",
                KnowledgeCategory.CRYPTOGRAPHY, listOf("crypto_basics"), "crack", "epic", "Run 'crack' on a hash"),
            KnowledgeFragment("defense_basics", "Defense Foundations", "Hardening systems against attack",
                KnowledgeCategory.DEFENSE, listOf(), "firewall", "uncommon", "Run 'firewall' command"),
            KnowledgeFragment("stealth_basics", "Stealth Tactics", "Avoiding detection",
                KnowledgeCategory.STEALTH, listOf(), "spoof", "uncommon", "Run 'spoof' command"),
            KnowledgeFragment("stealth_advanced", "Ghost Protocol", "Become invisible on the wire",
                KnowledgeCategory.STEALTH, listOf("stealth_basics"), "proxy", "rare", "Chain 'proxy' for anonymity"),
            KnowledgeFragment("network_basics", "Network Fundamentals", "Understanding the network layer",
                KnowledgeCategory.NETWORKING, listOf(), "ifconfig", "common", "Run 'ifconfig'"),
            KnowledgeFragment("persistence", "Persistence Techniques", "Maintaining access to a system",
                KnowledgeCategory.EXPLOITATION, listOf("exploit_basics"), "backdoor", "epic", "Run 'backdoor' on a target"),
            KnowledgeFragment("propagation", "Worm Propagation", "Self-replicating payloads",
                KnowledgeCategory.EXPLOITATION, listOf("persistence"), "worm", "legendary", "Run 'worm' with --spread"),
            KnowledgeFragment("ai_basics", "AI Assistance", "Leveraging AI for hacking",
                KnowledgeCategory.GENERAL, listOf("crypto_advanced", "stealth_advanced"), "ai-assist", "legendary", "Run 'ai-assist'")
        )
    }
}

data class KnowledgeMapData(
    val totalFragments: Int,
    val discoveredFragments: Int,
    val unlockedCommands: Int,
    val totalCommands: Int,
    val fragments: List<FragmentStatus>
)

data class FragmentStatus(
    val id: String,
    val name: String,
    val description: String,
    val category: String,
    val rarity: String,
    val unlocked: Boolean,
    val available: Boolean,
    val requiredFragments: List<String>,
    val unlocksCommand: String?,
    val sourceHint: String
)

File: zero-server-csharp/src/ZeroDayMMO.Server/Services/HackToLearnService.cs
using ZeroDayMMO.Shared.Models;

namespace ZeroDayMMO.Server.Services;

public enum KnowledgeCategory
{
    RECONNAISSANCE, EXPLOITATION, CRYPTOGRAPHY,
    DEFENSE, STEALTH, NETWORKING, GENERAL
}

public class KnowledgeFragment
{
    public string Id { get; set; } = "";
    public string Name { get; set; } = "";
    public string Description { get; set; } = "";
    public KnowledgeCategory Category { get; set; }
    public List<string> RequiredFragments { get; set; } = new();
    public string? UnlocksCommand { get; set; }
    public string Rarity { get; set; } = "common";
    public string SourceHint { get; set; } = "";
}

public record KnowledgeMapData(
    int TotalFragments,
    int DiscoveredFragments,
    int UnlockedCommands,
    int TotalCommands,
    List<FragmentStatus> Fragments
);

public record FragmentStatus(
    string Id,
    string Name,
    string Description,
    string Category,
    string Rarity,
    bool Unlocked,
    bool Available,
    List<string> RequiredFragments,
    string? UnlocksCommand,
    string SourceHint
);

public class HackToLearnService
{
    private readonly IPlayerService _playerService;
    private readonly GameEventBus? _gameEventBus;
    private readonly Dictionary<string, HashSet<string>> _playerKnowledge = new();
    private readonly Dictionary<string, HashSet<string>> _playerDiscoveredCommands = new();
    private readonly SemaphoreSlim _mutex = new(1, 1);

    public HackToLearnService(IPlayerService playerService, GameEventBus? gameEventBus = null)
    {
        _playerService = playerService;
        _gameEventBus = gameEventBus;
    }

    public async Task InitializePlayer(string playerId)
    {
        await _mutex.WaitAsync();
        try
        {
            _playerKnowledge.TryAdd(playerId, new HashSet<string>());
            _playerDiscoveredCommands.TryAdd(playerId, new HashSet<string>());
        }
        finally { _mutex.Release(); }
    }

    public async Task<bool> LearnKnowledge(string playerId, string fragmentId)
    {
        await _mutex.WaitAsync();
        try
        {
            var knowledge = _playerKnowledge.GetValueOrDefault(playerId);
            if (knowledge is null)
            {
                knowledge = new HashSet<string>();
                _playerKnowledge[playerId] = knowledge;
            }
            return knowledge.Add(fragmentId);
        }
        finally { _mutex.Release(); }
    }

    public async Task<HashSet<string>> GetPlayerKnowledge(string playerId)
    {
        await _mutex.WaitAsync();
        try
        {
            return _playerKnowledge.GetValueOrDefault(playerId)?.ToHashSet() ?? new HashSet<string>();
        }
        finally { _mutex.Release(); }
    }

    public async Task<KnowledgeFragment> DiscoverFragment(string playerId, string fragmentId)
    {
        await _mutex.WaitAsync();
        try
        {
            var fragment = KnowledgeFragments.Find(f => f.Id == fragmentId)
                ?? throw new InvalidOperationException("Knowledge fragment not found");

            var known = _playerKnowledge.GetValueOrDefault(playerId);
            if (known is null)
            {
                known = new HashSet<string>();
                _playerKnowledge[playerId] = known;
            }

            if (known.Contains(fragmentId))
                throw new InvalidOperationException("Already discovered this fragment");

            if (!fragment.RequiredFragments.All(r => known.Contains(r)))
                throw new InvalidOperationException($"Required fragments: {string.Join(", ", fragment.RequiredFragments)}");

            known.Add(fragmentId);

            if (fragment.UnlocksCommand is not null)
            {
                var discoveredCmds = _playerDiscoveredCommands.GetValueOrDefault(playerId);
                if (discoveredCmds is null)
                {
                    discoveredCmds = new HashSet<string>();
                    _playerDiscoveredCommands[playerId] = discoveredCmds;
                }
                discoveredCmds.Add(fragment.UnlocksCommand);

                var player = _playerService.GetPlayer(playerId);
                if (player is not null)
                {
                    player.UnlockedCommands.Add(fragment.UnlocksCommand);
                    _playerService.UpdatePlayer(player);
                }
            }

            _gameEventBus?.Publish(new GameEvent { PlayerId = playerId, Type = AchievementEvent.FRAGMENT_GATHERED, Value = 1 });
            return fragment;
        }
        finally { _mutex.Release(); }
    }

    public async Task<List<KnowledgeFragment>> GetAvailableFragments(string playerId)
    {
        await _mutex.WaitAsync();
        try
        {
            var known = _playerKnowledge.GetValueOrDefault(playerId) ?? new HashSet<string>();
            return KnowledgeFragments
                .Where(f => !known.Contains(f.Id) && f.RequiredFragments.All(r => known.Contains(r)))
                .ToList();
        }
        finally { _mutex.Release(); }
    }

    public async Task<KnowledgeMapData> GetKnowledgeMap(string playerId)
    {
        await _mutex.WaitAsync();
        try
        {
            var known = _playerKnowledge.GetValueOrDefault(playerId) ?? new HashSet<string>();
            var discoveredCmds = _playerDiscoveredCommands.GetValueOrDefault(playerId) ?? new HashSet<string>();

            return new KnowledgeMapData(
                KnowledgeFragments.Count,
                known.Count,
                discoveredCmds.Count,
                CommandRegistrySize,
                KnowledgeFragments.Select(f => new FragmentStatus(
                    f.Id, f.Name, f.Description,
                    f.Category.ToString(), f.Rarity,
                    Unlocked: known.Contains(f.Id),
                    Available: !known.Contains(f.Id) && f.RequiredFragments.All(r => known.Contains(r)),
                    f.RequiredFragments,
                    f.UnlocksCommand,
                    known.Contains(f.Id) ? "" : f.SourceHint
                )).ToList()
            );
        }
        finally { _mutex.Release(); }
    }

    private static int CommandRegistrySize { get; } = 100;

    private static readonly List<KnowledgeFragment> KnowledgeFragments = new()
    {
        new KnowledgeFragment { Id = "recon_basics", Name = "Recon Basics", Description = "Foundational reconnaissance techniques", Category = KnowledgeCategory.RECONNAISSANCE, RequiredFragments = new(), UnlocksCommand = "ping", Rarity = "common", SourceHint = "Run 'ping' or 'scan' to discover this" },
        new KnowledgeFragment { Id = "recon_advanced", Name = "Advanced Recon", Description = "Deep network probing and enumeration", Category = KnowledgeCategory.RECONNAISSANCE, RequiredFragments = new() { "recon_basics" }, UnlocksCommand = "nmap", Rarity = "uncommon", SourceHint = "Run 'nmap' on a target" },
        new KnowledgeFragment { Id = "exploit_basics", Name = "Exploit Fundamentals", Description = "Basic exploitation techniques", Category = KnowledgeCategory.EXPLOITATION, RequiredFragments = new() { "recon_advanced" }, UnlocksCommand = "exploit", Rarity = "rare", SourceHint = "Run 'exploit' on a target" },
        new KnowledgeFragment { Id = "crypto_basics", Name = "Crypto Foundations", Description = "Understanding encryption and hashing", Category = KnowledgeCategory.CRYPTOGRAPHY, RequiredFragments = new(), UnlocksCommand = "decrypt", Rarity = "common", SourceHint = "Run 'decrypt' on a file" },
        new KnowledgeFragment { Id = "crypto_advanced", Name = "Crypto Mastery", Description = "Breaking strong encryption", Category = KnowledgeCategory.CRYPTOGRAPHY, RequiredFragments = new() { "crypto_basics" }, UnlocksCommand = "crack", Rarity = "epic", SourceHint = "Run 'crack' on a hash" },
        new KnowledgeFragment { Id = "defense_basics", Name = "Defense Foundations", Description = "Hardening systems against attack", Category = KnowledgeCategory.DEFENSE, RequiredFragments = new(), UnlocksCommand = "firewall", Rarity = "uncommon", SourceHint = "Run 'firewall' command" },
        new KnowledgeFragment { Id = "stealth_basics", Name = "Stealth Tactics", Description = "Avoiding detection", Category = KnowledgeCategory.STEALTH, RequiredFragments = new(), UnlocksCommand = "spoof", Rarity = "uncommon", SourceHint = "Run 'spoof' command" },
        new KnowledgeFragment { Id = "stealth_advanced", Name = "Ghost Protocol", Description = "Become invisible on the wire", Category = KnowledgeCategory.STEALTH, RequiredFragments = new() { "stealth_basics" }, UnlocksCommand = "proxy", Rarity = "rare", SourceHint = "Chain 'proxy' for anonymity" },
        new KnowledgeFragment { Id = "network_basics", Name = "Network Fundamentals", Description = "Understanding the network layer", Category = KnowledgeCategory.NETWORKING, RequiredFragments = new(), UnlocksCommand = "ifconfig", Rarity = "common", SourceHint = "Run 'ifconfig'" },
        new KnowledgeFragment { Id = "persistence", Name = "Persistence Techniques", Description = "Maintaining access to a system", Category = KnowledgeCategory.EXPLOITATION, RequiredFragments = new() { "exploit_basics" }, UnlocksCommand = "backdoor", Rarity = "epic", SourceHint = "Run 'backdoor' on a target" },
        new KnowledgeFragment { Id = "propagation", Name = "Worm Propagation", Description = "Self-replicating payloads", Category = KnowledgeCategory.EXPLOITATION, RequiredFragments = new() { "persistence" }, UnlocksCommand = "worm", Rarity = "legendary", SourceHint = "Run 'worm' with --spread" },
        new KnowledgeFragment { Id = "ai_basics", Name = "AI Assistance", Description = "Leveraging AI for hacking", Category = KnowledgeCategory.GENERAL, RequiredFragments = new() { "crypto_advanced", "stealth_advanced" }, UnlocksCommand = "ai-assist", Rarity = "legendary", SourceHint = "Run 'ai-assist'" }
    };
}
